Read and Write arrays of nullable value types

Brar opened the pull request:
After deciding that my extended testing project should not hold this one up forever, I'm finally submitting this one that fixes #443 

I've explicitly broken backwards compatibility for `NpgsqlDataReader.GetFieldValue<T>()` where reading an array containing null values resulted in an array where the null values were replaced with the value types default value.

Another thing to note is that `NpgsqlDataReader.GetValue()` now returns an array of the value type or an array of its nullable counterpart on a row to row base, depending on the question whether the actual array returned by the backend contains null values or not.

roji commented:
Hey @Brar, thanks for the PR.

I'll give this a proper review, but right off the bat:

> Another thing to note is that NpgsqlDataReader.GetValue() now returns an array of the value type or an array of its nullable counterpart on a row to row base, depending on the question whether the actual array returned by the backend contains null values or not.

If I understand this correctly, this seems problematic to me. I don't think that the type of the value returned should depend on its contents in this way: it shifts a lot of responsibility on the user and will make it very easy to write buggy code. In effect, it means that whenever a row is read, it's necessary to to check what type was returned. IIRC it would also be the only case where CLR types read change across rows.

Let's revisit [our conversation on this](https://github.com/npgsql/npgsql/issues/443#issuecomment-347109462). Here's how I see things
* As a general rule, users are encouraged to use `reader.GetFieldValue<T>()`). 
* A user should be able to call `reader.GetFieldValue<int?[]>()`, yielding an array of nullables.
* A user should also be able to call `reader.GetFieldValue<int[]>()`, yielding an array of ints. If there are any nulls in the database column, this should throw an exception. I think this makes sense because the user has clearly expressed their intent for a non-nullable array, so in effect they're asserting that there will not be any nulls.
* The only open question in my mind is what `reader.GetValue()` returns. Here we have two options:
** Return `int[]`, silently returning zeros instead of nulls. The only advantage of this seems to be that it's the current behavior and therefore doesn't break backwards compat. However, as array use is somewhat rare I think we can afford to break backwards compat to do the right thing.
** Return `int[]`, throwing if nulls are present (just like `GetFieldValue<int[]>()`). The advantage of this is that it seems to correspond to the usual case of storing non-nullable int arrays in the database (unfounded intuition). The downside (as you mentioned earlier) is that `GetValue()` cannot be used to read some perfectly valid values.
** Return `int?[]` (just like `GetFieldValue<int?[]>()`). The main downside is that it's a bit more complicated to access the more usual case of non-nullable ints (we have those extra nullables). But I think of the three possibilities, this is the right thing to do.

What do you think? @YohDeadfall you may be interested as well.

Brar commented:
> If I understand this correctly, this seems problematic to me. I don't think that the type of the value returned should depend on its contents in this way: it shifts a lot of responsibility on the user and will make it very easy to write buggy code. In effect, it means that whenever a row is read, it's necessary to to check what type was returned.

I certainly considered this when thinking about the new interface.

My arguments are the following:

Basically an API is about promises guaranteed by the framework.

- If you (the user) want to make assumptions about the returned value: use `GetFieldValue<T>()` and we promise you to give you an object of the type you requested (as  `GetFieldValue<T>()` guarantees) or back out with an exception if it isn't compatible.
- If you (the user) don't want to make assumptions about the returned value (possibly because you don't know the exact table structure): use `GetValue()` and we promise you to give you an object (as  `GetValue()` guarantees) that provides the best compatibility to the type returned by the backend. You can even use `GetFieldType()` to find out that it is an `Array` type and cast it safely to Array. Afterwards you can use the `Array` instance methods to find out more about the actual array and cast it safely once more without even having to fall back to reflection.
- If you (the user) want to make assumptions about the returned value and want to use `GetValue()` you should be wary that the assumptions you are making when casting the returned object to something else are correct. After all, it's your data that contains the null values and the interface gives you everything you need to check if your assumptions are correct on a row to row base.

This also solves an important problem in our previous discussion:
In most cases, arrays will not contain null values.
Should all users of `GetValue()` have to pay the price of converting an Nullable<T>[] to T[] even if they know that their database will never have any null value inside an array?
I don't think so.

Finally:
I'm certainly willing to give this fact some documentation love.

roji commented:
I agree with the above, but returning a different type for different rows of the same database type seems to go too far.

> In most cases, arrays will not contain null values.
> Should all users of GetValue() have to pay the price of converting an Nullable<T>[] to T[] even if they know that their database will never have any null value inside an array?
> I don't think so.

It's true that we have this unverified intuition. However, I simply don't see the downside of returning `T?[]`, i.e. the price you're mentioning. If the user knows their database will never have any null value, then they should call `GetFieldValue<int?[]>()` and not `GetValue()`. If we're in some sort of dynamic situation where we don't know the type in advance, then I think it's safe to say that we don't know whether it contains nulls or not.

Also, I very much agree with what you wrote above, that with `GetValue()` "we promise you to give you an object (as GetValue() guarantees) that provides the best compatibility to the type returned by the backend". If we follow that to the letter that means we need to return an array of nullables, but that really is what is represented by PostgreSQL arrays. Note that we provide the best compatibility to the *type* returned by the backend, not to the *value*.

To summarize, I can see many more downsides to switching returned data types for different rows of the same resultset, than I can see for simply always returning `int?[]`.

Brar commented:
> If we follow that to the letter that means we need to return an array of nullables: that really is what is represented by PostgreSQL arrays. Note that we provide the best compatibility to the type returned by the backend, not to the value.

> To summarize, I can see many more downsides to switching returned data types for different rows of the same resultset, than I can see for simply always returning int?[].

I can agree with what you wrote @roji .

I'll modify the code accordingly.

Brar commented:
@roji the latest commit changes the implementation according to your comments above (I've squashed the others but left this one to make the changes transparent).

Brar commented:
Wow!
The number of failed test demonstrably proves the fact that this is a breaking change!
Will fix them.

roji commented:
Thanks @Brar, drop me a line when you want this reviewed.

Brar commented:
... but the fixes are quite easy.

@roji ready to review if you want.

Brar commented:
BTW: I've verified that merging with #1993 currently works automatically so there's no need to worry about which one to apply first, but I'll certainly do a rebase once #1993 is applied.

roji commented:
> One ugly part that we didn't talk about yet is the fact that ArrayHandler.ReadAsObject() is not only used by NpgsqlDataReader.GetValue() but also by NpgsqlCommand.ExecuteScalar() (see my fix for the failing range test) where the user doesn't have a way to ask for a specific type via generic method.

You're absolutely right. I think it makes it even more important to return a single, clear type and not alternate between `int[]` and `int?[]` based on whether there are nulls in the particular array.

>  We might add a generic method to NpgsqlCommand to give people a way to get a non-nullable array that, but it would be non-standard for current ADO.NET.

Good idea. I opened https://github.com/dotnet/corefx/issues/30426 for this, but we can also implement it as an Npgsql-specific feature in the meantime. Feel free to submit a PR on this if you want.

Brar commented:
@roji after working around all the issues I mentioned at https://github.com/dotnet/corefx/issues/30426 the code would look like this:

```csharp
public bool TryExecuteScalar<T>(out T result)
{
    var (hasResult, res) = TryExecuteScalar<T>(false, CancellationToken.None).GetAwaiter().GetResult();
    result = res;
    return hasResult;
}


[MethodImpl(MethodImplOptions.AggressiveInlining)]
[ItemCanBeNull]
async ValueTask<(bool hasResult, T result)> TryExecuteScalar<T>(bool async, CancellationToken cancellationToken)
{
    var behavior = CommandBehavior.SingleRow;
    if (!Parameters.HasOutputParameters)
        behavior |= CommandBehavior.SequentialAccess;
    using (var reader = await ExecuteDbDataReader(behavior, async, cancellationToken))
    {
        return reader.Read() && reader.FieldCount != 0 ? (true, reader.GetFieldValue<T>(0)) : (false, default);
    }
}
```
To be honest this doesn't look like an API to me.

Brar commented:
@roji I'm uncertain If you recognized the following a few days ago:

> @roji ready to review if you want.

roji commented:
I'll try to get around to this as soon as I can, although I have some very intensive non-Npgsql days coming up. I'll definitely get there eventually!

Brar commented:
Rebased to avoid code drift.
AppVeyor failure is not related to this PR (some sort of dependency issues with GeoAPI.CoordinateSystems).

YohDeadfall commented:
@Brar The AppVeyor issue was fixed by #2059.

Brar commented:
I've added a benchmark to measure the effect of boxing when reading value type arrays containing null values.
It certainly has a measurable effect but overall it seems acceptable to me.
I've copied the `Array` and `List\<T>` benchmarks for Arrays with and without `null`values into two separate tables but keep in mind that they come from different benchmarks so there is a certain margin of error.
Percent values in the config column are % of null values in the array.

As you will probably will spot, the `List\<T>` reading implementation needs some tuning which I will commit after this (replace reflection with Expressions/Delegate).

``` ini

BenchmarkDotNet=v0.11.5, OS=Windows 10.0.17763.437 (1809/October2018Update/Redstone5)
AMD Ryzen 5 2600X, 1 CPU, 12 logical and 6 physical cores
.NET Core SDK=3.0.100-preview4-011223
  [Host]     : .NET Core 3.0.0-preview4-27615-11 (CoreCLR 4.6.27615.73, CoreFX 4.700.19.21213), 64bit RyuJIT
  DefaultJob : .NET Core 3.0.0-preview4-27615-11 (CoreCLR 4.6.27615.73, CoreFX 4.700.19.21213), 64bit RyuJIT


```
|          Method | Config |            Mean |           Error |          StdDev |
|---------------- |------- |----------------:|----------------:|----------------:|
|    **ReadIntArray** |      **0** |        **261.5 ns** |       **0.6259 ns** |       **0.5227 ns** |
| ReadStringArray |      0 |        274.4 ns |       1.4087 ns |       1.3177 ns |
| **ReadNullableIntArray** |           **0** |        **271.9 ns** |       **0.8750 ns** |       **0.6831 ns** |
|      ReadStringArray |           0 |        281.1 ns |       0.5500 ns |       0.5144 ns |
|    **ReadIntArray** |      **1** |        **401.3 ns** |       **0.4550 ns** |       **0.3800 ns** |
| ReadStringArray |      1 |        582.3 ns |       1.8500 ns |       1.7305 ns |
| **ReadNullableIntArray** |        **1,0%** |        **729.7 ns** |       **5.0041 ns** |       **4.4360 ns** |
|      ReadStringArray |        1,0% |        581.8 ns |       1.6030 ns |       1.4995 ns |
| **ReadNullableIntArray** |      **1,100%** |        **654.6 ns** |       **2.4539 ns** |       **2.1753 ns** |
|      ReadStringArray |      1,100% |        407.2 ns |       0.8495 ns |       0.7946 ns |
|    **ReadIntArray** |     **10** |      **1,356.3 ns** |       **5.1931 ns** |       **4.8576 ns** |
| ReadStringArray |     10 |      2,585.4 ns |       7.0298 ns |       6.5757 ns |
| **ReadNullableIntArray** |       **10,0%** |      **2,063.1 ns** |       **8.0278 ns** |       **7.1164 ns** |
|      ReadStringArray |       10,0% |      2,676.7 ns |      17.6374 ns |      16.4980 ns |
| **ReadNullableIntArray** |      **10,33%** |      **1,942.8 ns** |       **5.9017 ns** |       **5.2317 ns** |
|      ReadStringArray |      10,33% |      2,289.7 ns |       3.9519 ns |       3.6966 ns |
| **ReadNullableIntArray** |      **10,50%** |      **1,790.6 ns** |       **8.9714 ns** |       **8.3918 ns** |
|      ReadStringArray |      10,50% |      1,957.6 ns |       4.3465 ns |       4.0657 ns |
| **ReadNullableIntArray** |      **10,66%** |      **1,783.5 ns** |       **6.1718 ns** |       **5.4711 ns** |
|      ReadStringArray |      10,66% |      1,859.9 ns |       7.8809 ns |       7.3718 ns |
| **ReadNullableIntArray** |     **10,100%** |      **1,362.7 ns** |       **3.8771 ns** |       **3.0270 ns** |
|      ReadStringArray |     10,100% |      1,154.1 ns |       5.5972 ns |       5.2356 ns |
|    **ReadIntArray** |   **1000** |    **104,584.9 ns** |     **491.3012 ns** |     **435.5257 ns** |
| ReadStringArray |   1000 |    222,209.4 ns |     401.5564 ns |     313.5088 ns |
| **ReadNullableIntArray** |     **1000,0%** |    **149,085.3 ns** |     **507.1815 ns** |     **423.5196 ns** |
|      ReadStringArray |     1000,0% |    223,018.6 ns |     507.9739 ns |     450.3057 ns |
| **ReadNullableIntArray** |    **1000,11%** |    **144,154.3 ns** |     **874.4626 ns** |     **730.2158 ns** |
|      ReadStringArray |    1000,11% |    212,397.5 ns |     818.1764 ns |     765.3227 ns |
| **ReadNullableIntArray** |    **1000,33%** |    **133,787.2 ns** |     **448.1872 ns** |     **349.9151 ns** |
|      ReadStringArray |    1000,33% |    181,575.2 ns |     589.1222 ns |     522.2415 ns |
| **ReadNullableIntArray** |    **1000,50%** |    **120,832.3 ns** |   **2,164.6832 ns** |   **2,024.8460 ns** |
|      ReadStringArray |    1000,50% |    163,224.2 ns |     503.5807 ns |     471.0497 ns |
| **ReadNullableIntArray** |    **1000,66%** |    **112,964.4 ns** |     **259.0522 ns** |     **242.3176 ns** |
|      ReadStringArray |    1000,66% |    135,243.0 ns |     325.0426 ns |     288.1418 ns |
| **ReadNullableIntArray** |    **1000,88%** |     **94,119.8 ns** |     **187.7104 ns** |     **175.5844 ns** |
|      ReadStringArray |    1000,88% |     98,104.4 ns |     216.7282 ns |     202.7277 ns |
| **ReadNullableIntArray** |   **1000,100%** |     **81,106.2 ns** |     **392.1711 ns** |     **366.8371 ns** |
|      ReadStringArray |   1000,100% |     83,420.2 ns |     114.5426 ns |     107.1433 ns |
|    **ReadIntArray** | **100000** | **10,372,510.0 ns** |  **29,219.8379 ns** |  **24,399.8880 ns** |
| ReadStringArray | 100000 | 55,528,871.4 ns | 194,028.7877 ns | 172,001.4664 ns |
| **ReadNullableIntArray** |   **100000,0%** | **14,547,172.1 ns** |  **25,350.8745 ns** |  **23,713.2237 ns** |
|      ReadStringArray |   100000,0% | 55,826,170.0 ns | 168,518.2639 ns | 157,632.0886 ns |
| **ReadNullableIntArray** |  **100000,11%** | **13,999,944.3 ns** |  **62,291.5218 ns** |  **58,267.5281 ns** |
|      ReadStringArray |  100000,11% | 52,086,735.3 ns | 290,923.5537 ns | 272,130.0726 ns |
| **ReadNullableIntArray** |  **100000,33%** | **12,993,367.7 ns** |  **31,636.3317 ns** |  **29,592.6443 ns** |
|      ReadStringArray |  100000,33% | 41,678,262.7 ns | 239,298.3322 ns | 199,824.9453 ns |
| **ReadNullableIntArray** |  **100000,50%** | **11,259,409.7 ns** |  **32,792.4079 ns** |  **30,674.0387 ns** |
|      ReadStringArray |  100000,50% | 33,310,327.5 ns | 150,792.6505 ns | 141,051.5389 ns |
| **ReadNullableIntArray** |  **100000,66%** | **10,805,462.4 ns** |  **29,094.5588 ns** |  **27,215.0684 ns** |
|      ReadStringArray |  100000,66% | 25,678,164.2 ns |  65,566.2796 ns |  61,330.7386 ns |
| **ReadNullableIntArray** |  **100000,88%** |  **8,853,761.9 ns** |  **33,249.8149 ns** |  **31,101.8975 ns** |
|      ReadStringArray |  100000,88% | 13,801,175.4 ns |  50,597.6919 ns |  47,329.1124 ns |
| **ReadNullableIntArray** | **100000,100%** |  **8,307,270.7 ns** |  **18,172.6085 ns** |  **16,998.6693 ns** |
|      ReadStringArray | 100000,100% |  8,097,250.6 ns |  81,465.1403 ns |  76,202.5428 ns |

``` ini

BenchmarkDotNet=v0.11.5, OS=Windows 10.0.17763.437 (1809/October2018Update/Redstone5)
AMD Ryzen 5 2600X, 1 CPU, 12 logical and 6 physical cores
.NET Core SDK=3.0.100-preview4-011223
  [Host]     : .NET Core 3.0.0-preview4-27615-11 (CoreCLR 4.6.27615.73, CoreFX 4.700.19.21213), 64bit RyuJIT
  DefaultJob : .NET Core 3.0.0-preview4-27615-11 (CoreCLR 4.6.27615.73, CoreFX 4.700.19.21213), 64bit RyuJIT


```
|           Method | Config |            Mean |           Error |          StdDev |
|----------------- |------- |----------------:|----------------:|----------------:|
|    **ReadListOfInt** |      **0** |        **210.8 ns** |       **0.9351 ns** |       **0.8747 ns** |
| ReadListOfString |      0 |        248.2 ns |       1.1680 ns |       1.0925 ns |
| **ReadListOfNullableInt** |           **0** |      **1,416.7 ns** |       **2.9541 ns** |       **2.6188 ns** |
|      ReadListOfString |           0 |        250.0 ns |       0.5292 ns |       0.4691 ns |
|    **ReadListOfInt** |      **1** |        **344.5 ns** |       **0.3864 ns** |       **0.3425 ns** |
| ReadListOfString |      1 |        531.3 ns |       1.3271 ns |       1.2414 ns |
| **ReadListOfNullableInt** |        **1,0%** |      **2,593.2 ns** |       **5.7867 ns** |       **5.4129 ns** |
|      ReadListOfString |        1,0% |        529.2 ns |       0.8649 ns |       0.7667 ns |
| **ReadListOfNullableInt** |      **1,100%** |      **2,581.9 ns** |      **11.3716 ns** |      **10.6370 ns** |
|      ReadListOfString |      1,100% |        423.0 ns |       1.3159 ns |       1.2309 ns |
|    **ReadListOfInt** |     **10** |      **1,315.1 ns** |       **2.4468 ns** |       **2.1691 ns** |
| ReadListOfString |     10 |      2,605.1 ns |       5.1842 ns |       4.8493 ns |
| **ReadListOfNullableInt** |       **10,0%** |      **3,774.4 ns** |       **6.9425 ns** |       **6.4940 ns** |
|      ReadListOfString |       10,0% |      2,590.0 ns |       7.5949 ns |       7.1042 ns |
| **ReadListOfNullableInt** |      **10,33%** |      **3,764.6 ns** |       **4.7224 ns** |       **3.9434 ns** |
|      ReadListOfString |      10,33% |      2,280.1 ns |       3.1685 ns |       2.9638 ns |
| **ReadListOfNullableInt** |      **10,50%** |      **3,475.4 ns** |       **4.4058 ns** |       **4.1212 ns** |
|      ReadListOfString |      10,50% |      2,052.4 ns |       3.0591 ns |       2.7118 ns |
| **ReadListOfNullableInt** |      **10,66%** |      **3,490.1 ns** |       **7.6208 ns** |       **7.1285 ns** |
|      ReadListOfString |      10,66% |      1,896.8 ns |       2.1963 ns |       1.9470 ns |
| **ReadListOfNullableInt** |     **10,100%** |      **3,071.8 ns** |       **7.2743 ns** |       **6.8044 ns** |
|      ReadListOfString |     10,100% |      1,203.7 ns |       3.0904 ns |       2.7396 ns |
|    **ReadListOfInt** |   **1000** |    **106,381.2 ns** |      **58.3635 ns** |      **45.5664 ns** |
| ReadListOfString |   1000 |    230,806.0 ns |     477.9400 ns |     447.0654 ns |
| **ReadListOfNullableInt** |     **1000,0%** |    **129,655.6 ns** |     **110.6112 ns** |      **98.0539 ns** |
|      ReadListOfString |     1000,0% |    225,867.1 ns |     529.3027 ns |     469.2131 ns |
| **ReadListOfNullableInt** |    **1000,11%** |    **125,629.8 ns** |     **414.0194 ns** |     **367.0174 ns** |
|      ReadListOfString |    1000,11% |    216,373.6 ns |     419.2856 ns |     392.2000 ns |
| **ReadListOfNullableInt** |    **1000,33%** |    **115,362.1 ns** |     **162.9168 ns** |     **144.4215 ns** |
|      ReadListOfString |    1000,33% |    186,132.7 ns |     507.5601 ns |     474.7720 ns |
| **ReadListOfNullableInt** |    **1000,50%** |     **97,335.7 ns** |     **135.6428 ns** |     **126.8804 ns** |
|      ReadListOfString |    1000,50% |    169,999.9 ns |     301.6281 ns |     282.1431 ns |
| **ReadListOfNullableInt** |    **1000,66%** |     **95,501.4 ns** |     **464.2364 ns** |     **434.2470 ns** |
|      ReadListOfString |    1000,66% |    138,287.6 ns |     317.4466 ns |     296.9398 ns |
| **ReadListOfNullableInt** |    **1000,88%** |     **74,684.4 ns** |     **132.8532 ns** |     **124.2710 ns** |
|      ReadListOfString |    1000,88% |    103,345.2 ns |     145.1422 ns |     128.6648 ns |
| **ReadListOfNullableInt** |   **1000,100%** |     **63,626.8 ns** |      **76.4018 ns** |      **71.4663 ns** |
|      ReadListOfString |   1000,100% |     83,881.4 ns |     268.6718 ns |     251.3158 ns |
|    **ReadListOfInt** | **100000** | **10,536,575.0 ns** |  **24,174.7365 ns** |  **22,613.0635 ns** |
| ReadListOfString | 100000 | 56,023,628.9 ns | 232,558.7336 ns | 217,535.5837 ns |
| **ReadListOfNullableInt** |   **100000,0%** | **12,773,653.6 ns** |  **26,859.7601 ns** |  **25,124.6363 ns** |
|      ReadListOfString |   100000,0% | 55,758,963.0 ns | 213,091.0041 ns | 199,325.4574 ns |
| **ReadListOfNullableInt** |  **100000,11%** | **12,234,889.6 ns** |  **19,569.1693 ns** |  **18,305.0131 ns** |
|      ReadListOfString |  100000,11% | 52,202,240.0 ns | 202,771.9332 ns | 189,672.9921 ns |
| **ReadListOfNullableInt** |  **100000,33%** | **11,483,795.3 ns** |  **31,479.7278 ns** |  **27,905.9588 ns** |
|      ReadListOfString |  100000,33% | 41,800,594.4 ns | 154,926.2986 ns | 144,918.1558 ns |
| **ReadListOfNullableInt** |  **100000,50%** |  **9,493,378.3 ns** |  **18,907.7713 ns** |  **17,686.3410 ns** |
|      ReadListOfString |  100000,50% | 33,079,586.7 ns |  76,097.3004 ns |  71,181.4620 ns |
| **ReadListOfNullableInt** |  **100000,66%** |  **9,028,969.6 ns** |  **30,967.4360 ns** |  **27,451.8254 ns** |
|      ReadListOfString |  100000,66% | 25,726,945.1 ns |  43,292.3207 ns |  38,377.5146 ns |
| **ReadListOfNullableInt** |  **100000,88%** |  **6,998,723.1 ns** |  **15,852.7410 ns** |  **14,828.6638 ns** |
|      ReadListOfString |  100000,88% | 13,940,472.7 ns |  56,131.2753 ns |  52,505.2297 ns |
| **ReadListOfNullableInt** | **100000,100%** |  **6,069,082.7 ns** |  **28,932.2599 ns** |  **27,063.2539 ns** |
|      ReadListOfString | 100000,100% |  8,593,221.0 ns |  51,163.5359 ns |  47,858.4033 ns |

Brar commented:
After optimization:

``` ini

BenchmarkDotNet=v0.11.5, OS=Windows 10.0.17763.437 (1809/October2018Update/Redstone5)
AMD Ryzen 5 2600X, 1 CPU, 12 logical and 6 physical cores
.NET Core SDK=3.0.100-preview4-011223
  [Host]     : .NET Core 3.0.0-preview4-27615-11 (CoreCLR 4.6.27615.73, CoreFX 4.700.19.21213), 64bit RyuJIT
  DefaultJob : .NET Core 3.0.0-preview4-27615-11 (CoreCLR 4.6.27615.73, CoreFX 4.700.19.21213), 64bit RyuJIT


```
|                Method |      Config |            Mean |           Error |          StdDev |
|---------------------- |------------ |----------------:|----------------:|----------------:|
| **ReadListOfNullableInt** |           **0** |        **227.1 ns** |       **1.5445 ns** |       **1.3692 ns** |
|      ReadListOfString |           0 |        248.3 ns |       0.3331 ns |       0.3116 ns |
| **ReadListOfNullableInt** |        **1,0%** |        **410.5 ns** |       **0.5906 ns** |       **0.5235 ns** |
|      ReadListOfString |        1,0% |        522.0 ns |       2.2086 ns |       2.0659 ns |
| **ReadListOfNullableInt** |      **1,100%** |        **321.6 ns** |       **0.4102 ns** |       **0.3837 ns** |
|      ReadListOfString |      1,100% |        417.9 ns |       1.3134 ns |       1.2286 ns |
| **ReadListOfNullableInt** |       **10,0%** |      **1,563.6 ns** |       **8.7317 ns** |       **8.1676 ns** |
|      ReadListOfString |       10,0% |      2,583.4 ns |      20.6839 ns |      19.3478 ns |
| **ReadListOfNullableInt** |      **10,33%** |      **1,434.8 ns** |       **3.1125 ns** |       **2.7592 ns** |
|      ReadListOfString |      10,33% |      2,323.7 ns |       4.7093 ns |       3.9325 ns |
| **ReadListOfNullableInt** |      **10,50%** |      **1,257.9 ns** |       **4.0342 ns** |       **3.7736 ns** |
|      ReadListOfString |      10,50% |      2,099.6 ns |       8.7580 ns |       7.3133 ns |
| **ReadListOfNullableInt** |      **10,66%** |      **1,268.3 ns** |       **4.5549 ns** |       **4.2607 ns** |
|      ReadListOfString |      10,66% |      1,880.0 ns |       4.4212 ns |       4.1356 ns |
| **ReadListOfNullableInt** |     **10,100%** |        **932.2 ns** |       **2.1852 ns** |       **1.9371 ns** |
|      ReadListOfString |     10,100% |      1,204.9 ns |       3.0922 ns |       2.8925 ns |
| **ReadListOfNullableInt** |     **1000,0%** |    **129,804.6 ns** |     **648.0870 ns** |     **606.2210 ns** |
|      ReadListOfString |     1000,0% |    226,472.5 ns |   1,132.6564 ns |   1,004.0704 ns |
| **ReadListOfNullableInt** |    **1000,11%** |    **123,673.3 ns** |   **1,029.3984 ns** |     **912.5348 ns** |
|      ReadListOfString |    1000,11% |    217,028.3 ns |   1,057.4190 ns |     989.1104 ns |
| **ReadListOfNullableInt** |    **1000,33%** |    **112,647.6 ns** |     **543.9553 ns** |     **454.2273 ns** |
|      ReadListOfString |    1000,33% |    186,034.1 ns |     627.3884 ns |     523.8977 ns |
| **ReadListOfNullableInt** |    **1000,50%** |     **96,163.8 ns** |     **410.5791 ns** |     **384.0559 ns** |
|      ReadListOfString |    1000,50% |    169,157.1 ns |     524.0043 ns |     490.1540 ns |
| **ReadListOfNullableInt** |    **1000,66%** |     **90,394.5 ns** |     **124.5299 ns** |     **110.3925 ns** |
|      ReadListOfString |    1000,66% |    137,948.5 ns |     248.2284 ns |     220.0480 ns |
| **ReadListOfNullableInt** |    **1000,88%** |     **71,423.6 ns** |     **182.6348 ns** |     **161.9010 ns** |
|      ReadListOfString |    1000,88% |    103,498.4 ns |     562.0904 ns |     525.7797 ns |
| **ReadListOfNullableInt** |   **1000,100%** |     **66,616.4 ns** |     **228.8309 ns** |     **214.0486 ns** |
|      ReadListOfString |   1000,100% |     83,209.4 ns |     158.6926 ns |     132.5155 ns |
| **ReadListOfNullableInt** |   **100000,0%** | **12,970,724.5 ns** |  **65,969.7769 ns** |  **61,708.1703 ns** |
|      ReadListOfString |   100000,0% | 55,647,906.8 ns |  76,996.7307 ns |  64,295.7574 ns |
| **ReadListOfNullableInt** |  **100000,11%** | **12,210,254.1 ns** |  **32,070.0482 ns** |  **28,429.2624 ns** |
|      ReadListOfString |  100000,11% | 53,119,874.0 ns | 205,991.7445 ns | 192,684.8056 ns |
| **ReadListOfNullableInt** |  **100000,33%** | **11,396,185.2 ns** |  **41,602.5952 ns** |  **38,915.0934 ns** |
|      ReadListOfString |  100000,33% | 41,663,050.3 ns | 243,009.9104 ns | 227,311.6209 ns |
| **ReadListOfNullableInt** |  **100000,50%** |  **9,466,906.4 ns** |  **16,982.5482 ns** |  **15,885.4861 ns** |
|      ReadListOfString |  100000,50% | 33,378,230.8 ns | 138,591.4390 ns | 129,638.5181 ns |
| **ReadListOfNullableInt** |  **100000,66%** |  **8,835,563.0 ns** |  **27,842.7698 ns** |  **26,044.1441 ns** |
|      ReadListOfString |  100000,66% | 25,880,455.8 ns |  51,358.8386 ns |  48,041.0895 ns |
| **ReadListOfNullableInt** |  **100000,88%** |  **6,965,653.1 ns** |   **9,147.3357 ns** |   **8,556.4235 ns** |
|      ReadListOfString |  100000,88% | 14,005,509.2 ns |  29,901.9424 ns |  26,507.2931 ns |
| **ReadListOfNullableInt** | **100000,100%** |  **6,549,614.4 ns** |  **11,721.6677 ns** |  **10,964.4552 ns** |
|      ReadListOfString | 100000,100% |  8,671,885.8 ns |  57,505.9005 ns |  53,791.0550 ns |

roji commented:
@Brar it will probably take me a bit to get around to looking at this again in the very near future - I'd also recommend holding off a bit until the big nullability PR is merged, so that you don't get too many conflicts afterwards...

Brar commented:
Certainly - no problem!
I was just thinking about doing a benchmark after our last communication regarding the boxing and when I found the performance issue with the nullable list I couldn't resist to fix it.

roji commented:
Oh I see, of course :)

Brar commented:
After a lot of discussions, reviews and rewriting I now consider this ready for a serious final review.
I've left two commits to reflect the last evolution of the PR.
The first commit reflects my initial work on this, which implemented the full set of functionality.
The second commit reflects the changes @roji and I did in a joint pair programming effort plus some minor changes from me where we left over some side issues.

As @roji already knows most of this, I explicitly request a review from @austindrenski and @YohDeadfall in order to get another couple of eyes to look at it and probably give some completely new feedback.

Below you can find another two benchmark runs.
The first one is without the changes from this PR, the second one is with the changes from this PR.

<details><summary>Click to see the benchmark results</summary>
<p>

**Interpretation hints:**

We have four methods

1. `[BeforePR_]ReadIntArray` which calls `NpgsqlDataReader.GetFieldValue<int[]>()` before and after the PR
2. `[BeforePR_]ReadNullableIntArray` which calls `NpgsqlDataReader.GetFieldValue<int[]>()` before, and `NpgsqlDataReader.GetFieldValue<int?[]>()` after the PR.
3. `[BeforePR_]ReadStringArray` which calls `NpgsqlDataReader.GetFieldValue<string[]>()` before and after the PR

When `AllNulls` is `false` the array contains only actual `int` values otherwise it only contains `null` values, except for `ReadIntArray` where the array never contains any `null` values (`AllNulls` is ignored).
`NumElements` is the number of elements in the array.

``` ini

BenchmarkDotNet=v0.11.5, OS=Windows 10.0.18362
AMD Ryzen 5 2600X, 1 CPU, 12 logical and 6 physical cores
.NET Core SDK=3.0.100
  [Host]     : .NET Core 3.0.0 (CoreCLR 4.700.19.46205, CoreFX 4.700.19.46214), 64bit RyuJIT
  DefaultJob : .NET Core 3.0.0 (CoreCLR 4.700.19.46205, CoreFX 4.700.19.46214), 64bit RyuJIT

```

|                        Method | AllNulls | NumElements |            Mean |           Error |          StdDev |
|------------------------------ |--------- |------------ |----------------:|----------------:|----------------:|
|         BeforePR_ReadIntArray |    False |           1 |        410.6 ns |       0.9145 ns |       0.8107 ns |
| BeforePR_ReadNullableIntArray |    False |           1 |        410.2 ns |       3.2328 ns |       3.0240 ns |
|      BeforePR_ReadStringArray |    False |           1 |        561.1 ns |       2.9994 ns |       2.8057 ns |
|         BeforePR_ReadIntArray |    False |          10 |      1,367.0 ns |       3.8014 ns |       3.5558 ns |
| BeforePR_ReadNullableIntArray |    False |          10 |      1,400.3 ns |       5.9320 ns |       5.5488 ns |
|      BeforePR_ReadStringArray |    False |          10 |      2,084.8 ns |       4.0968 ns |       3.8322 ns |
|         BeforePR_ReadIntArray |    False |        1000 |    101,709.1 ns |     133.8482 ns |     118.6530 ns |
| BeforePR_ReadNullableIntArray |    False |        1000 |    101,319.7 ns |      78.1266 ns |      65.2392 ns |
|      BeforePR_ReadStringArray |    False |        1000 |    167,909.3 ns |     778.4015 ns |     690.0327 ns |
|         BeforePR_ReadIntArray |    False |      100000 | 10,403,245.7 ns |  48,560.5826 ns |  45,423.5991 ns |
| BeforePR_ReadNullableIntArray |    False |      100000 | 10,451,426.3 ns |  80,794.1426 ns |  71,621.9030 ns |
|      BeforePR_ReadStringArray |    False |      100000 | 27,716,898.3 ns |  73,197.6963 ns |  68,469.1705 ns |
|         BeforePR_ReadIntArray |     True |           1 |        430.2 ns |      10.0778 ns |      14.1277 ns |
| BeforePR_ReadNullableIntArray |     True |           1 |        351.6 ns |       7.0207 ns |       6.8952 ns |
|      BeforePR_ReadStringArray |     True |           1 |        446.2 ns |       5.9172 ns |       5.5350 ns |
|         BeforePR_ReadIntArray |     True |          10 |      1,425.1 ns |      16.9358 ns |      15.0132 ns |
| BeforePR_ReadNullableIntArray |     True |          10 |        809.9 ns |      12.6105 ns |      11.7958 ns |
|      BeforePR_ReadStringArray |     True |          10 |      1,181.2 ns |       8.6565 ns |       8.0973 ns |
|         BeforePR_ReadIntArray |     True |        1000 |    102,055.6 ns |     153.8075 ns |     143.8716 ns |
| BeforePR_ReadNullableIntArray |     True |        1000 |     47,903.4 ns |     180.8129 ns |     169.1325 ns |
|      BeforePR_ReadStringArray |     True |        1000 |     79,895.0 ns |      66.0380 ns |      58.5410 ns |
|         BeforePR_ReadIntArray |     True |      100000 | 10,419,988.3 ns | 134,581.9563 ns | 125,888.0455 ns |
| BeforePR_ReadNullableIntArray |     True |      100000 |  4,867,496.4 ns |  15,627.9646 ns |  14,618.4078 ns |
|      BeforePR_ReadStringArray |     True |      100000 |  8,151,878.0 ns |  17,476.1494 ns |  16,347.2010 ns |


|               Method | AllNulls | NumElements |            Mean |           Error |          StdDev |
|--------------------- |--------- |------------ |----------------:|----------------:|----------------:|
|         ReadIntArray |    False |           1 |        358.9 ns |       1.1590 ns |       1.0842 ns |
| ReadNullableIntArray |    False |           1 |        438.2 ns |       1.5376 ns |       1.4382 ns |
|      ReadStringArray |    False |           1 |        484.3 ns |       0.8036 ns |       0.7517 ns |
|         ReadIntArray |    False |          10 |      1,284.3 ns |       1.4725 ns |       1.2296 ns |
| ReadNullableIntArray |    False |          10 |      2,043.4 ns |       8.0207 ns |       7.5026 ns |
|      ReadStringArray |    False |          10 |      2,268.8 ns |       5.2173 ns |       4.8803 ns |
|         ReadIntArray |    False |        1000 |    103,898.2 ns |     492.2491 ns |     460.4501 ns |
| ReadNullableIntArray |    False |        1000 |    176,518.5 ns |     838.7399 ns |     743.5211 ns |
|      ReadStringArray |    False |        1000 |    199,062.7 ns |     532.9325 ns |     498.5054 ns |
|         ReadIntArray |    False |      100000 | 10,512,378.0 ns |  29,216.6596 ns |  22,810.4460 ns |
| ReadNullableIntArray |    False |      100000 | 17,722,589.4 ns |  50,355.3856 ns |  47,102.4589 ns |
|      ReadStringArray |    False |      100000 | 29,388,754.4 ns | 147,610.4754 ns | 138,074.9302 ns |
|         ReadIntArray |     True |           1 |        356.7 ns |       0.8660 ns |       0.7232 ns |
| ReadNullableIntArray |     True |           1 |        303.6 ns |       1.2909 ns |       1.1444 ns |
|      ReadStringArray |     True |           1 |        350.3 ns |       0.6431 ns |       0.5701 ns |
|         ReadIntArray |     True |          10 |      1,299.8 ns |       3.5177 ns |       3.2905 ns |
| ReadNullableIntArray |     True |          10 |        934.6 ns |       2.5992 ns |       2.4313 ns |
|      ReadStringArray |     True |          10 |      1,135.1 ns |       2.2193 ns |       2.0759 ns |
|         ReadIntArray |     True |        1000 |    108,434.3 ns |      80.1156 ns |      66.9002 ns |
| ReadNullableIntArray |     True |        1000 |     71,092.4 ns |     788.4061 ns |     698.9014 ns |
|      ReadStringArray |     True |        1000 |     87,305.6 ns |     174.1468 ns |     154.3766 ns |
|         ReadIntArray |     True |      100000 | 10,516,978.6 ns |  58,233.5149 ns |  54,471.6660 ns |
| ReadNullableIntArray |     True |      100000 |  7,027,959.4 ns |  10,050.7037 ns |   8,392.7928 ns |
|      ReadStringArray |     True |      100000 |  8,958,207.8 ns |  16,675.6081 ns |  15,598.3742 ns |

</p>
</details>

Brar commented on src/Npgsql/TypeHandlers/ArrayHandler.cs:
```diff
@@ -23,85 +25,69 @@ namespace Npgsql.TypeHandlers
         {
-            if (IsArrayOf<TAny, TElement>.Value)
-                return (TAny)(object)await ReadArray<TElement>(buf, async);
+            Debug.Assert(ArrayTypeInfo<TRequestedArray>.IsArrayOrList);
```
You're probably right.
I'd consider the exception as vital, because `ArrayHandler` is on the public API surface, so this method might be called in a way that we don't expect or support but in this case the `Debug.Assert` is really not necessary.

I think it's a leftover from some phase during refactoring when the method body was more complicated and we felt like we'd need it to make clear what the method expects.
I'll remove it.

Brar commented on src/Npgsql/TypeHandlers/ArrayHandler.cs:
```diff
@@ -150,31 +135,159 @@ protected async ValueTask<Array> ReadArray<TAnyElement>(NpgsqlReadBuffer buf, bo
+            var containsNulls = buf.ReadInt32() == 1;
+            if (!ElementTypeInfo<TRequestedElement>.IsReferenceOrNullable && containsNulls)
+                throw new InvalidOperationException(
+                    $"Can't read a non-nullable List<{typeof(TRequestedElement).Name}> if the database array field contains null values. Use List<{typeof(TRequestedElement).Name}?> instead.");
```
Ok, fixed

Brar commented on src/Npgsql/TypeHandlers/ArrayHandler.cs:
```diff
@@ -150,31 +135,159 @@ protected async ValueTask<Array> ReadArray<TAnyElement>(NpgsqlReadBuffer buf, bo
+                var asyncParam = Expression.Parameter(typeof(bool), "async");
+
+                ReadNullableArrayFunc = Expression.Lambda<Func<ArrayHandler, NpgsqlReadBuffer, bool, ValueTask<Array>>>(
+                        Expression.Call(
```
Ok

Brar commented on src/Npgsql/TypeHandlers/ArrayHandler.cs:
```diff
@@ -150,31 +135,159 @@ protected async ValueTask<Array> ReadArray<TAnyElement>(NpgsqlReadBuffer buf, bo
+                if (IsArray)
+                {
+                    ReadArrayFunc = Expression.Lambda<Func<ArrayHandler, NpgsqlReadBuffer, bool, ValueTask<Array>>>(
+                            Expression.Call(
```
Ok.

Brar commented on src/Npgsql/TypeHandlers/ArrayHandler.cs:
```diff
@@ -150,31 +135,159 @@ protected async ValueTask<Array> ReadArray<TAnyElement>(NpgsqlReadBuffer buf, bo
+            public static readonly List<TElement> EmptyList = new List<TElement>(0);
+            // ReSharper disable StaticMemberInGenericType
+            public static readonly Func<ArrayHandler, NpgsqlReadBuffer, bool, ValueTask<Array>> ReadNullableArrayFunc = default!;
+            // ReSharper restore StaticMemberInGenericType
```
No, IMO it doesn't.
It warns us, that we'll end up with one instance of `ReadNullableArrayFunc`  per `TElement` which is exactly what we want as we use `ElementTypeInfo<TElement>` as cache for the compiled expression trees. Mind you, that `ElementTypeInfo<TElement>` is nested inside `ArrayHandler` not inside `ArrayHandler<TElement>` so in this case `TElement` might be the element type of any array that the consumer requested.
Perhaps it might be a good idea to rename `TElement` to `TAnyElement` in this case to avoid confusion.
What do you think?

YohDeadfall replied:
Ah, yes, you are right (:

Brar commented on src/Npgsql/TypeHandlers/ArrayHandler.cs:
```diff
@@ -150,31 +135,159 @@ protected async ValueTask<Array> ReadArray<TAnyElement>(NpgsqlReadBuffer buf, bo
+                                typeof(Nullable<>).MakeGenericType(typeof(TElement))),
+                            bufferParam, asyncParam),
+                        arrayHandlerParam, bufferParam, asyncParam)
+                    .Compile();
```
That's not possible, because, as I wrote above, `ElementTypeInfo<TElement>` is nested inside `ArrayHandler` not inside `ArrayHandler<TElement>` so in this case `TElement` might be the element type of any array that the consumer requested.

Brar replied:
> so in this case `TElement` might be the element type of any array that the consumer requested.

It's even worse, as in this case `TElement` isn't even available as we're inside `ArrayHandler` not inside `ArrayHandler<TElement>` which is intentional as we want to support both `TElement` and `TElement?`.

YohDeadfall replied:
Yep, my mistake. There is a little difference between methods I lost, but hen there is a question. What if an array must have no nulls in it because its elements are `NOT NULL` domains? Right now we a going to force users to use array with nullable elements, this is a breakage. At the same time it will be published in v5.0, so it's okay probably.

@austindrenski @roji, your thoughts?

YohDeadfall replied:
What I asked first is possible, but it's too tricky and introduce more code than your solution. Let's keep it as is in that case (:

Brar replied:
Good question.
I've actually never thought about domains with a `NOT NULL` constraint as a way to avoid `NULL` values inside an array.
While trying to look up if this is even supported I found out that the use of `NOT NULL` constraints on domains is [mildly discouraged](https://www.postgresql.org/docs/current/sql-createdomain.html#id-1.9.3.62.7) because it has some tricky corner cases but I've confirmed that it is possible to create and use arrays of non-nullable domains, which renders your question perfectly valid.
On the other hand we're probably talking about a corner case of a corner case feature and we only force them if they use `NpgsqlDataReader.GetValue()`. If they use `NpgsqlDataReader.GetFieldValue<T>()` they can always request an array of non-nullable value types.

roji replied:
I really like the idea of `NOT NULL` domains - good idea @YohDeadfall. Even if `NOT NULL` has its caveats, it's very clear information on the type expressing the intention for it never to contain nulls.

I agree with @Brar that it's a corner case, but I still do think it would be nice to do, at least for completeness. It also shouldn't be very hard: [pg_type contains `typnotnull`](https://www.postgresql.org/docs/current/catalog-pg-type.html) which could be loaded and stored on PostgresDomainType, and we could check that from ArrayHandler.GetValue.

But we shouldn't block this PR on this (unless you really feel like doing this right away @Brar). Unless someone objects I can open a separate issue to traack this.

YohDeadfall replied:
I'm okay with doing this in the next iteration, @Brar.

Brar replied:
I'm all for the separate issue.
I also feel like it's not too complicated, so we can possibly get this done for 5.0 but I really want to avoid feature creep for this PR as it has already taken long enough to bring it to a state that we all can possibly agree on.
I'll have a look at it after we've merged this, if you want me to.

roji replied:
Opened #2720 to track, @Brar you're welcome to take this if you're interested.

Brar commented:
Thank you for the review @YohDeadfall.
I've addressed your feedback where it was possible but there are two unresolved comments left where I think that it's not possible to do what you suggested.
Can you please look at those once again?

YohDeadfall commented on src/Npgsql/TypeHandlers/ArrayHandler.cs:
```diff
@@ -150,31 +133,164 @@ protected async ValueTask<Array> ReadArray<TAnyElement>(NpgsqlReadBuffer buf, bo
             return list;
         }
 
+        static InvalidOperationException UnsupportedNonNullableTypeException(string requestedType, string suggestedType)
```
It would be better to pass types instead of strings.

Brar replied:
Please look at the call sites.
Do you really think it's worth to do the complex type-construction-fu that would be necessary to pass the correct type just for type safety of an internal method.
Or am I missing something obvious?

YohDeadfall replied:
https://github.com/npgsql/npgsql/blob/43919c0b72a7c2ad7852146df5c3f451d165fbac/src/Npgsql/TypeHandlers/ArrayHandler.cs#L85

As you can see the suggested array type has only one dimension while there could be requested a  multidimensional array. It's incorrect, so you can change exception message to omit collection type and say "Cannot read a non-nullable collection of elements because the returned array contains nulls". Otherwise, create an array first and use `GetType`.

Brar replied:
You're right. I think I'll go with the simplified exception message.

YohDeadfall commented on src/Npgsql/TypeHandlers/ArrayHandler.cs:
```diff
@@ -112,24 +96,23 @@ protected async ValueTask<Array> ReadArray<TAnyElement>(NpgsqlReadBuffer buf, bo
             {
-                var element = await _elementHandler.ReadWithLength<TAnyElement>(buf, async);
+                var element = await ElementHandler.ReadWithLength<TRequestedElement>(buf, async);
                 result.SetValue(element, indices);
```
Are you interested in optimizing this in the second PR? There is not so hard, just allocate a default boxed value before the loop in the case of value type, and use `Unsafe.Unbox<T>` to get the reference to the value and replace it each iteration passing the same boxed value to `SetValue`.

Brar replied:
That's a good Idea but I think that it's more complicated than you think as `Unsafe.Unbox<T>` has a `struct` type constraint which would mean that we have to use expression trees once again.
TBH I'd leave this for a follow-up PR.

YohDeadfall replied:
Take a look at `NullableHandler` which uses some kind of magic to read and write nullable values (: By the way, you can use the same trick instead of code generation via expressions.

Brar replied:
Ok, I gave it a closer look and immediately ran into the following problem:

> CS8177 C# Async methods cannot have by-reference locals

I think that for now there are enough obstacles to justify postponing this for a later PR.

YohDeadfall commented on src/Npgsql/TypeHandlers/ArrayHandler.cs:
```diff
@@ -150,31 +133,164 @@ protected async ValueTask<Array> ReadArray<TAnyElement>(NpgsqlReadBuffer buf, bo
+        {
+            public static readonly bool IsReferenceOrNullable =
+                !typeof(TElement).IsValueType ||
+                typeof(TElement).IsGenericType && typeof(TElement).GetGenericTypeDefinition() == typeof(Nullable<>);
```
You can invert this by renaming the field to `IsNonNullable` or `NotNullble` (as you wish) and changing expression to `typeof(TElement).IsValueType && Nullable.GetUnderlyingType(typeof(TElement)) is null`.

Brar replied:
Good suggestion!
I'll change it.

Brar commented:
Thank you @YohDeadfall 

@austindrenski and @roji, @YohDeadfall and I have completed the review process. Now it's on you.

roji commented:
Great work @Brar, and thanks @YohDeadfall for the deep review and helpful comments. Very happy this is finally merged :)

christophertrml commented:
Hi npgsql team!

Would you be able to kick off a build of the 5.0.0 unstable branch that includes this merge now that it's merged? Would love to try it out.

Thanks!

roji commented:
Hmm, looks like our appveyor build has some issue and doesn't upload unstable packages anymore... Rather than spending time to debug this, hopefully we'll merge #2717 soon (switch to Github Actions). @christophertrml for now you can clone and build Npgsql (shouldn't be hard).

Brar commented:
> Hmm, looks like our appveyor build has some issue and doesn't upload unstable packages anymore...

I couldn't resist investigating this...

See #2722
